# Thông tin:
- Họ tên: Nguyễn Phúc Thuần
- MSSV: 20120380

# Bài làm

## 3.23

In [28]:
from math import sqrt

def isPerfectSquare(x):
    root = sqrt(x)
    return pow(int(root + 0.5), 2) == x

def factorize(n):
    b = 1
    
    while True:
        if b == n:
            return (None, None)
        
        a = n + pow(b,2)
        
        print(f"{n} + {b}^2 = {a}")
        
        if isPerfectSquare(a):
            root = int(sqrt(a))
            
            first_factor = root + b
            second_factor = root - b
            
            return (first_factor, second_factor)
        
        b = b + 1

In [29]:
# a
print(factorize(53357))

# b
print(factorize(34571))

#c
print(factorize(25777))

#d 
print(factorize(64213))

53357 + 1^2 = 53358
53357 + 2^2 = 53361
(233, 229)
34571 + 1^2 = 34572
34571 + 2^2 = 34575
34571 + 3^2 = 34580
34571 + 4^2 = 34587
34571 + 5^2 = 34596
(191, 181)
25777 + 1^2 = 25778
25777 + 2^2 = 25781
25777 + 3^2 = 25786
25777 + 4^2 = 25793
25777 + 5^2 = 25802
25777 + 6^2 = 25813
25777 + 7^2 = 25826
25777 + 8^2 = 25841
25777 + 9^2 = 25858
25777 + 10^2 = 25877
25777 + 11^2 = 25898
25777 + 12^2 = 25921
(173, 149)
64213 + 1^2 = 64214
64213 + 2^2 = 64217
64213 + 3^2 = 64222
64213 + 4^2 = 64229
64213 + 5^2 = 64238
64213 + 6^2 = 64249
64213 + 7^2 = 64262
64213 + 8^2 = 64277
64213 + 9^2 = 64294
64213 + 10^2 = 64313
64213 + 11^2 = 64334
64213 + 12^2 = 64357
64213 + 13^2 = 64382
64213 + 14^2 = 64409
64213 + 15^2 = 64438
64213 + 16^2 = 64469
64213 + 17^2 = 64502
64213 + 18^2 = 64537
64213 + 19^2 = 64574
64213 + 20^2 = 64613
64213 + 21^2 = 64654
64213 + 22^2 = 64697
64213 + 23^2 = 64742
64213 + 24^2 = 64789
64213 + 25^2 = 64838
64213 + 26^2 = 64889
64213 + 27^2 = 64942
64213 + 28^2 = 64997
64213

## 3.25
### (a) N = 61063
$1882^2 \cdot 1898^2 \equiv (2\cdot 3^3 \cdot 5)(2\cdot3^5\cdot5^3) \text{ mod N}$

$= (2^2\cdot3^8\cdot5^4) = (2\cdot3^4\cdot5^2)^2 = 4050^2$

$\Rightarrow gcd(61063, 1882\cdot1898 - 4050) = 227$

$\Rightarrow 61063 = 227\cdot269$

### (b) N = 52907
$763^2\cdot773^2 \equiv 2^6\cdot3\cdot2^6\cdot3^5 \text{ mod N}$

$= (2^6\cdot3^3)^2 = 1728^2$

$\Rightarrow gcd(52907, 763\cdot773 - 1728) = 277$

$\Rightarrow 52907 = 277\cdot191$

### (c) N = 198103
$1189^2\cdot1605^2\cdot2815^2 \equiv (2^3\cdot3^3\cdot5^3)\cdot(2\cdot7^3)\cdot(3\cdot5\cdot7) \text{ mod N}$

$= (2^2\cdot3^2\cdot5^2\cdot7^2)^2=44100$

$\Rightarrow gcd(198103, 1189\cdot1605\cdot2815 - 44100) = 499$

$\Rightarrow 198103 = 499\cdot397$

### (d) N = 2525891
$1591^2\cdot5275^2\cdot5401^2\equiv(2\cdot5\cdot7^2\cdot11)\cdot(2^3\cdot3^6\cdot7)\cdot(2^4\cdot3^2\cdot5^3\cdot7\cdot 11)\text{ mod N}$

$=(2^4\cdot3^4\cdot5^2\cdot7^2\cdot11)^2 = 17463600^2$

$\Rightarrow gcd(N, 1591\cdot5275\cdot5401 - 17463600)=1637$

$\Rightarrow N = 1637\times1543$

## 3.29
### (a)
Xét giới hạn: $H = \lim\limits_{X \rightarrow \infty}{\dfrac{L(X)}{{ln(X)}^\alpha}}$
$=\lim\limits_{X \rightarrow \infty}{\dfrac{e^{\sqrt{lnX \cdot lnlnX}}}{{ln(X)}^\alpha}}$

$\lim\limits_{X \rightarrow \infty}e^{ln{{\dfrac{e^{\sqrt{lnX \cdot lnlnX}}}{{ln(X)}^\alpha}}}}$

$= \lim\limits_{X \rightarrow \infty}{e^{ln{e^{\sqrt{lnX \cdot lnlnX}} - ln{ln(X)}^\alpha}}}$

$= \lim\limits_{X \rightarrow \infty}{e^{\sqrt{lnX \cdot lnlnX} - ln{ln(X)}^\alpha}} \rightarrow e^\infty$

Dễ thấy khi $X \rightarrow \infty$ thì $H > 0$

Do đó ta nói $L(X) = \Omega{(ln(X)^\alpha)}$


### (b)
Xét giới hạn: $H = \lim\limits_{X \rightarrow \infty}{\dfrac{L(X)}{X^\beta}}$
$=\lim\limits_{X \rightarrow \infty}{\dfrac{e^{\sqrt{lnX \cdot lnlnX}}}{X^\beta}}$

$=\lim\limits_{X \rightarrow \infty}{e^{ln{\dfrac{e^{\sqrt{lnX \cdot lnlnX}}}{X^\beta}}}}$

$=\lim\limits_{X \rightarrow \infty}{e^{\sqrt{lnX \cdot lnlnX} -ln{X^\beta}}} \rightarrow e^{-\infty}$

Dễ thấy khi $X \rightarrow \infty$ thì $H \rightarrow 0 < \infty$

Do đó ta nói: $L(X) = O(X^\beta)$

## 3.31
### (a)
Đầu tiên, để $(lnX)^\epsilon < (lnX)^{1- \epsilon}$ thì tiên quyết $\epsilon < 1 - \epsilon \Rightarrow \epsilon < \frac{1}{2}$

Thứ hai, $lnL(X) = ln(e^{\sqrt{lnX \cdot lnlnX}}) = \sqrt{lnX \cdot lnlnX} = \sqrt{lnlnX} \cdot lnX^{\frac{1}{2}}$

Xét thấy $\sqrt{lnlnX}$ khá nhỏ, chỉ đủ để làm lệch giá trị của $lnX^{\frac{1}{2}}$

mà $(lnX)^\epsilon < (lnX)^{\frac{1}{2}}< (lnX)^{1- \epsilon} \forall x < \frac{1}{2}$

Do đó chắc chắn $\forall \epsilon < \frac{1}{2}$ ta luôn có $(lnX)^\epsilon < \ln L(X)< (lnX)^{1- \epsilon}$

### (b)
Xét $u = \dfrac{\ln X}{\ln Y} = \dfrac{\ln X}{\ln L(X)^c} = \dfrac{\ln X}{c \cdot \sqrt{\ln X \cdot \ln\ln X}}$

$=\dfrac{1}{c} \cdot \sqrt\dfrac{\ln X}{\ln X \cdot \ln\ln X}$

## 3.34
### a,
$ u + v = -3 + 7β +2β^2 -2β^3 \\ \\
uv = -10 - 49β + 35β^2 + 42β^3 -158β^4 + 126β^5 - 63β^6 \\ \\
\text{Vì bậc của β chỉ lên đến bậc 3 nên: } \\ \\ 
\ chia \ cho \ f(β) = 1 + 3β -2β^3 + β^4 \ ta \ được \ \\ \\
uv = 148 + 425β + 98β^2 - 85β^3 ∈ Z[β].
$

### b,
$ u + v = 5 + 3β +4β^2 -10β^3 \\ \\
uv = -36 - 90β - 96β^2 - 79β^3 -27β^4 - 20β^5 + 25β^6 \\ \\
\text{Vì bậc của β chỉ lên đến bậc 3 nên: } \\ \\ 
\ chia \ cho \ f(β) = 1 + 3β -2β^3 + β^4 \ ta \ được \ \\ \\
uv = -69 - 219β - 211β^2 - 88β^3 ∈ Z[β].
$

### c,
$ u + v = 4 + 2β + 9β^2 + 3β^3 \\ \\
uv = -12 + 52β - 5β^2 - 15β^3 + 39β^4 + 18β^5  \\ \\
\text{Vì bậc của β chỉ lên đến bậc 3 nên: } \\ \\ 
\ chia \ cho \ f(β) = 1 + 3β -2β^3 + β^4 \ ta \ được \ \\ \\
uv = -87 - 191β - 59β^2 + 135β^3 ∈ Z[β].
$

## 3.35
Ta có: $p = 19079$, $g = 17$

$g^{3030} ≡ 14580  \text{ (mod p)}$

mà $14580 = 2^2 \cdot 3^6 \cdot 5 $

Suy ra $g^i$ với $i = 3030$ là  5-smooth

$g^{6892} ≡ 18432  \text{ (mod p)}$

mà $18432 = 2^{11} \cdot 3^2$

Suy ra $g^i$ với $i = 6892$ là  5-smooth

$g^{18312} ≡ 6000  \text{ (mod p)}$

mà $6000 = 2^4 \cdot 3 \cdot 5^3$

Suy ra $g^i$ với $i = 18312$ là  5-smooth 

## 3.36

### a)
${a^{\frac{p-1}{2}}}^2 = a^{p-1} ≡ 1 \text{ (mod p)}$ 

Như vậy $ {a^{\frac{p-1}{2}}}^2$ là một nghiệm đồng dư $x^2 ≡ 1 \text{ (mod p)}$

Phương trình đồng dư trên có hai nghiệm là $x = 1$ và $x = -1$.

Như vậy $ {a^{\frac{p-1}{2}}}^2 ≡ ±1 \text{ (mod p)}$

### b)
Ta có ${a^{\frac{p-1}{2}}}^2 ≡ ±1 \text{ (mod p)}$ nên $a^2 ≡ 2^{p−1} \text{ (mod p)}$

Theo định lí Fermat nhỏ ta ta có:

$ a^{p-1} \cdot (p-1)! ≡ (p-1)! \text{ (mod p)}$

$\Rightarrow a^{p-1} = a^p-a ≡ 1 \text{ (mod p)}$

Như vậy ta được $ a^2=2^{p−1} ≡ 1 \text{ (mod p)}$

## 3.38
 Vì $ p ≡ 3 \ (mod \ 4) \ \\ nên \ p = 4n + 3 \\ 
Suy \ \ ra \ \ \frac{p-1}{2} = 2n + 1 \ \ \  \text{ là số lẻ} $
Ta có
$ a^{\frac{p+1}{2}} = (a^{\frac{p+1}{4}})^2 \\
và \ a^{\frac{p+1}{2}} = a^{1+ \frac{p-1}{2}} = a.a^{\frac{p-1}{2}}\\
do \ đó \ b^2 ≡ (a^{\frac{p+1}{4}})^2  \ \ \ \ (mod \ p)  \ \ \ \ (1)
$

Áp dụng kết quả của bài tập 3.36b, ta có: 
$a^{\frac{p-1}{2}} ≡ 1  \ \ \ \ (mod \ p)  \\
Suy \ \ ra \ \ (a^{\frac{p+1}{4}})^2 = a^{\frac{p+1}{2}} ≡ a  \ \ \ \ (mod \ p) \ \ \ \ \ \ (2) $

Từ (1) và (2) suy ra 
$ b^2 ≡ a   \ \ \ \ (mod \ p) \ \ \ \ \  (đpcm) $